File: wpilibj/src/test/java/edu/wpi/first/wpilibj/experimental/command/DefaultCommandTest.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2018 FIRST. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package edu.wpi.first.wpilibj.experimental.command;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class DefaultCommandTest extends CommandTestBase {
  @Test
  void defaultCommandScheduleTest() {
    CommandScheduler scheduler = new CommandScheduler();

    TestSubsystem hasDefaultCommand = new TestSubsystem();

    MockCommandHolder defaultHolder = new MockCommandHolder(true, hasDefaultCommand);
    Command defaultCommand = defaultHolder.getMock();

    scheduler.setDefaultCommand(hasDefaultCommand, defaultCommand);
    scheduler.run();

    assertTrue(scheduler.isScheduled(defaultCommand));
  }

  @Test
  void defaultCommandInterruptResumeTest() {
    CommandScheduler scheduler = new CommandScheduler();

    TestSubsystem hasDefaultCommand = new TestSubsystem();

    MockCommandHolder defaultHolder = new MockCommandHolder(true, hasDefaultCommand);
    Command defaultCommand = defaultHolder.getMock();
    MockCommandHolder interrupterHolder = new MockCommandHolder(true, hasDefaultCommand);
    Command interrupter = interrupterHolder.getMock();

    scheduler.setDefaultCommand(hasDefaultCommand, defaultCommand);
    scheduler.run();
    scheduler.schedule(interrupter);

    assertFalse(scheduler.isScheduled(defaultCommand));
    assertTrue(scheduler.isScheduled(interrupter));

    scheduler.cancel(interrupter);
    scheduler.run();

    assertTrue(scheduler.isScheduled(defaultCommand));
    assertFalse(scheduler.isScheduled(interrupter));
  }
}

File: wpilibc/src/test/native/cpp/frc2/commands/DefaultCommandTest.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2019 FIRST. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include "CommandTestBase.h"
#include "frc/frc2/commands/RunCommand.h"

using namespace frc2;
class DefaultCommandTest : public CommandTestBase {

};

TEST_F(DefaultCommandTest, DefaultCommandScheduleTest) {
  CommandScheduler scheduler = GetScheduler();

  TestSubsystem subsystem;

  RunCommand command1([]{}, {&subsystem});

  scheduler.SetDefaultCommand(&subsystem, &command1);
  scheduler.Run();

  EXPECT_TRUE(scheduler.IsScheduled(&command1));
}

TEST_F(DefaultCommandTest, DefaultCommandInterruptResumeTest) {
  CommandScheduler scheduler = GetScheduler();

  TestSubsystem subsystem;

  RunCommand command1([]{}, {&subsystem});
  RunCommand command2([]{}, {&subsystem});

  scheduler.SetDefaultCommand(&subsystem, &command1);
  scheduler.Run();
  scheduler.Schedule(&command2);

  EXPECT_TRUE(scheduler.IsScheduled(&command2));
  EXPECT_FALSE(scheduler.IsScheduled(&command1));

  scheduler.Cancel(&command2);
  scheduler.Run();

  EXPECT_TRUE(scheduler.IsScheduled(&command1));
}
